lib/dockerregistry/inventory.go

```diff
@@ -255,6 +279,76 @@ func ParseManifestsFromDir(
 	return mfests, nil
 }
 
+// ValidateThinManifestDirectoryStructure enforces a particular directory
+// structure for thin manifests. Most importantly, it requires that if a file
+// named "foo/manifests/bar/promoter-manifest.yaml" exists, that a corresponding
+// file named "foo/images/bar/promoter-manifest.yaml" must also exist.
+//
+// nolint[gocyclo]
+func ValidateThinManifestDirectoryStructure(
+	dir string,
+) error {
+	// First, enforce that there are directories named "images" and "manifests".
+	dirExists, err := directoryExists(filepath.Join(dir, "images"))
+	if !dirExists || err != nil {
+		return err
+	}
+
+	mfestDir := filepath.Join(dir, "manifests")
+	dirExists, err = directoryExists(mfestDir)
+	if !dirExists || err != nil {
+		return err
+	}
+
+	// For every subfolder in <dir>/manifests, ensure that a
+	// "promoter-manifest.yaml" file exists, and also that a corresponding file
+	// exists in the "images" folder.
+	files, err := ioutil.ReadDir(mfestDir)
+	if err != nil {
+		return err
+	}
+	for _, file := range files {
+		p, err := os.Stat(filepath.Join(mfestDir, file.Name()))
+		if err != nil {
+			return err
+		}
+
+		// Only check for directory files.
+		if p.IsDir() {
+			// Check if "promoter-manifest.yaml" exists here.
+			if _, err := os.Stat(
+				filepath.Join(mfestDir,
+					file.Name(),
+					"promoter-manifest.yaml")); os.IsNotExist(err) {
+				continue
```

Comment:
Blocker (?): This doesn't match the comment - the comment says that promoter-manifest.yaml must exist, but this tolerates it not existing.  It looks like it actually tolerates _any_ error.